lib/Conversion/ExportVerilog/ExportVerilog.cpp

```diff
@@ -6026,6 +6027,79 @@ void ModuleEmitter::emitHWModule(HWModuleOp module) {
   currentModuleOp = nullptr;
 }
 
+//===----------------------------------------------------------------------===//
+// Emitter for files & file lists.
+//===----------------------------------------------------------------------===//
+
+class FileEmitter : public EmitterBase {
+public:
+  explicit FileEmitter(VerilogEmitterState &state) : EmitterBase(state) {}
+
+  void emit(emit::FileOp op);
+  void emit(emit::FileListOp op);
+
+private:
+  void emitOp(emit::VerbatimOp op);
+};
+
+void FileEmitter::emit(emit::FileOp op) {
+  for (Operation &op : *op.getBodyBlock()) {
+    TypeSwitch<Operation *, void>(&op)
+        .Case<emit::VerbatimOp>([&](auto op) { emitOp(op); })
+        .Default([&](auto op) {
+          op->emitError("cannot emit operation to file");
+          state.encounteredError = true;
+        });
+  }
+  ps.eof();
+}
+
+void FileEmitter::emit(emit::FileListOp op) {
+  // Find the associated file ops and write the paths on individual lines.
+  for (auto sym : op.getFiles()) {
+    auto fileName = cast<FlatSymbolRefAttr>(sym).getAttr();
+    auto file = cast<emit::FileOp>(state.symbolCache.getDefinition(fileName));
+    ps << PP::neverbox;
+    ps << PPExtString(file.getFileName());
+    ps << PP::end << PP::newline;
+  }
+  ps.eof();
+}
+
+void FileEmitter::emitOp(emit::VerbatimOp op) {
+  startStatement();
+
+  SmallPtrSet<Operation *, 8> ops;
+  ops.insert(op);
+  ps << PP::neverbox;
+
+  // Drop an extraneous \n off the end of the string if present.
+  StringRef string = op.getText();
+  if (string.ends_with("\n"))
+    string = string.drop_back();
+
+  // Emit each \n separated piece of the string with each piece properly
+  // indented.  The convention is to not emit the \n so
+  // emitLocationInfoAndNewLine can do that for the last line.
+  bool isFirst = true;
+
+  // Emit each line of the string at a time.
+  while (!string.empty()) {
+    auto lhsRhs = string.split('\n');
+    if (isFirst)
+      isFirst = false;
+    else {
+      ps << PP::end << PP::newline << PP::neverbox;
+    }
+    ps << PPExtString(lhsRhs.first);
+    string = lhsRhs.second;
+  }
+
```

Comment:
Can we avoid code duplication with SV verbatim op emission?

Reply:
Agree.